File: 238-product-of-array-except-self/238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        std::vector<int> ans(nums.size(),1);
    
        
       for(int i = 1; i < nums.size(); i++)
       {
           ans[i] = ans[i-1] * nums[i-1]; 
       }
        
        int product = 1;
        
       for(int i = nums.size()-2; i >= 0; i--){
           product*=nums[i+1];
           ans[i] *= product;
           
       }
        
        return ans; 
    }
};

File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        pre = []
        
        for _ in range(len(nums)):
            pre.append(1)
            
        for i in range(1, len(pre)):
            pre[i] = nums[i-1] *  pre[i-1]
        
        prev = nums[-1]
        for i in range(len(pre) -2, -1, -1):
            pre[i] *= prev
            prev *= nums[i]

        return pre
            
        
        
        
        